src/dbnode/storage/profiler/profiler.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package profiler contains the code used for profiling.
+package profiler
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime/pprof"
+	"sync/atomic"
+)
+
+const (
+	// ProfileFileExtension is the extension of profile files.
+	ProfileFileExtension = ".pb.gz"
+)
+
+// FileProfileContext is file profiler context data.
+type FileProfileContext struct {
+	path        string
+	profileName *profileName
+}
+
+// StopProfile stops started profile.
+func (f FileProfileContext) StopProfile() error {
+	stopCPUProfile()
+	return writeHeapProfile(f.path, f.profileName)
```

Comment:
I think heap dump should be it's own method and not coupled with the end of CPU profiling. E.g. one may want to take a heap dump before and after some process, or without doing any CPU profiling.